client/src/main.rs

```diff
@@ -374,26 +375,41 @@ fn main() {
                     let add_location_calls = spec.locations().into_iter().skip(1).map(|l| add_location_call(&api.metadata, cid, l)).collect();
                     let add_location_batch_call = batch_call(&api.metadata, add_location_calls);
 
-                    info!("raw 'new_community' sudo call to sign with js/apps {}: 0x{}",
-                        cid, hex::encode(sudo_call(&api.metadata, new_community_call.clone()).encode())
-                    );
-
-                    info!("raw sudo 'add_location' batch call to sign with js/apps {}: 0x{}",
-                        cid, hex::encode(sudo_call(&api.metadata.clone(), add_location_batch_call.clone()).encode())
-                    );
+                    if contains_sudo_pallet(&api.metadata) {
+                        info!("Printing raw sudo calls for js/apps for cid: {}", cid);
+                        print_raw_call("sudo(new_community)", &sudo_call(&api.metadata, new_community_call.clone()));
+                        print_raw_call("sudo(utility_batch(add_location))", &sudo_call(&api.metadata.clone(), add_location_batch_call.clone()));
 
-                    // ---- send xt's to chain
-                    send_and_wait_for_in_block(&api, sudo_xt(&api, new_community_call));
-                    println!("{}", cid);
+                        // ---- send xt's to chain
+                        send_and_wait_for_in_block(&api, sudo_xt(&api, new_community_call));
+                        println!("{}", cid);
 
-                    if api.get_current_phase().unwrap() != CeremonyPhaseType::REGISTERING {
-                        error!("Wrong ceremony phase for registering new locations for {}", cid);
-                        error!("Aborting without registering additional locations");
-                        std::process::exit(exit_code::WRONG_PHASE);
-                    }
+                        if api.get_current_phase().unwrap() != CeremonyPhaseType::REGISTERING {
+                            error!("Wrong ceremony phase for registering new locations for {}", cid);
+                            error!("Aborting without registering additional locations");
+                            std::process::exit(exit_code::WRONG_PHASE);
+                        }
+                        send_and_wait_for_in_block(&api, sudo_xt(&api, add_location_batch_call));
 
-                    send_and_wait_for_in_block(&api, sudo_xt(&api, add_location_batch_call));
+                    } else {
+                        info!("Printing raw collective propose calls for js/apps for cid: {}", cid);
```

Comment:
Quite some redundant code here. I think we could do the if only around the call wrapping and leave the rest the same for the two cases

Reply:
;-)

Reply:
If I understand you correctly. It is not as simple as you might think. In rust you can't do:

```
let call = if contains_sudo_pallet(&api.metadata) {
                        sudo_call(sudo_call(&api.metadata, new_community_call.clone())
                    } else {
                        collective_propose_call(&api.metadata, 1, new_community_call)
};
```

Because they are different types.

Reply:
Solved by introducing `OpaqueCall` type.